### PR TITLE
Added 2 helper functions for creating plaster manifest files. Updated…

### PR DESCRIPTION
**Pull request by zloeber**
- Added some minor updates to the xsd schema to include more parameters for new-modulemanifest.
- Added code to invoke-plaster to accommodate the change above (and automatically look for more module manifest settings to pull in for existing manifest files.
- Added 2 helper functions to more easily create both parameter and content xml code snippets for plaster manifest files from standard PowerShell hashes (Write-PlasterContent and Write-PlasterParameter).

### CODE REVIEW

**Comment by rkeithhill** [comment]
Thanks for the contribution!  This probably would have been better to submit as two PRs - one for the schema change and another for the addition of the two commands.  But we'll go with it combined for now.  I'll have a number of comments - mostly style but there are a few other non-style issues to address.

File: src/InvokePlaster.ps1 [hunk 1]
```diff
@@ -805,7 +805,8 @@ function Invoke-Plaster {
                     if ($? -and $oldModuleManifest) {
                         $props = 'Guid', 'Description', 'DefaultCommandPrefix', 'RootModule', 'AliasesToExport',
                                  'CmdletsToExport', 'DscResourcesToExport', 'VariablesToExport',
-                                 'FormatsToProcess', 'TypesToProcess', 'ScriptsToProcess'
+                                 'FormatsToProcess', 'TypesToProcess', 'ScriptsToProcess', 'Tags', 'Author',
```
Comment (rkeithhill) [inline src/InvokePlaster.ps1]: I don't mind adding these but I worry a bit about trying to stay up-to-date with new module manifest fields.  The original idea was to provide support for just the commonly modified fields provided as part of a template.  But what is "commonly modified fields" is definitely debatable.  :-)

File: src/InvokePlaster.ps1 [hunk 2]
```diff
@@ -826,6 +827,21 @@ function Invoke-Plaster {
                 if (![string]::IsNullOrWhiteSpace($description)) {
                     $newModuleManifestParams['Description'] = $description
                 }
+                if (![string]::IsNullOrWhiteSpace($copyright)) {
+                    $newModuleManifestParams['Copyright'] = $copyright
+                }
+                if (![string]::IsNullOrWhiteSpace($projectUri)) {
+                    $newModuleManifestParams['projectUri'] = $projectUri
```
Comment (rkeithhill) [inline src/InvokePlaster.ps1]: Not that it matters (PS being case-insensitive), I would PascalCase `projectUri` (to `ProjectUri`) along with the rest of the hashtable keys below.

File: src/Plaster.psd1 [hunk 3]
```diff
@@ -50,10 +50,12 @@
     # exported.  This improves performance of PowerShell when discovering the commands in
     # module.
     FunctionsToExport = @(
-        'Invoke-Plaster'
-        'New-PlasterManifest'
+        'Invoke-Plaster',
```
Comment (rkeithhill) [inline src/Plaster.psd1]: Adding the commas is an unnecessary change.

File: src/Schema/PlasterManifest-v1.xsd [hunk 4]
```diff
@@ -448,6 +448,31 @@
                       <xs:documentation>Specifies the value of the RootModule property.</xs:documentation>
                     </xs:annotation>
                   </xs:attribute>
+                  <xs:attribute name="copyright" type="xs:string">
```
Comment (rkeithhill) [inline src/Schema/PlasterManifest-v1.xsd]: Backwards compatible changes to the schema (like adding new attributes), requires bumping the SchemaVersion (to 1.2 in this case) on line 13 of this file and line 76 of Plaster.psm1.  This allows Plaster to alert users who are invoking a template based on the updated schema via an older Plaster module, that they will need to update Plaster in order to process the template.

File: src/Schema/PlasterManifest-v1.xsd [hunk 5]
```diff
@@ -448,6 +448,31 @@
...
+                    <xs:annotation>
+                      <xs:documentation>Specifies the copyright for this module.</xs:documentation>
+                    </xs:annotation>
+                  </xs:attribute>
+                  <xs:attribute name="tags" type="xs:string">
```
Comment (rkeithhill) [inline src/Schema/PlasterManifest-v1.xsd]: BTW camelCase here is fine. Thanks for following the convention in the schema.

File: src/WritePlasterContent.ps1 [hunk 6]
```diff
@@ -0,0 +1,246 @@
+function Write-PlasterContent {
+    <#
```
Comment (rkeithhill) [inline src/WritePlasterContent.ps1]: Please follow the help comment style of the rest of the project.  Comment help goes outside and above the function.

File: src/WritePlasterContent.ps1 [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+    .SYNOPSIS
+        A simple helper function to create a plaster content xml block
+    .DESCRIPTION
+        A simple helper function to create a plaster content xml block. Types of plaster content blocks
+        include: File, TemplateFile, Message, Modify, newModuleManifest, requireModule
+
+    .EXAMPLE
+        TBD
+    #>
+    [cmdletbinding()]
```
Comment (rkeithhill) [inline src/WritePlasterContent.ps1]: PascalCase e.g. `[CmdletBinding()]`

File: src/WritePlasterContent.ps1 [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    Param (
```
Comment (rkeithhill) [inline src/WritePlasterContent.ps1]: lower case here for `param` to be consistent with the rest of the code

File: src/WritePlasterContent.ps1 [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+        [Parameter(ParameterSetName = "default", Mandatory = $true, Position = 0)]
+        [ValidateSet('File','TemplateFile','Message','Modify','newModuleManifest','requireModule')]
```
Comment (rkeithhill) [inline src/WritePlasterContent.ps1]: Let's be consistent here and use PascalCase for all values.

File: src/WritePlasterContent.ps1 [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+        [string]$RequiredVersion,
+
+        [Parameter(ParameterSetName = "default")]
+        [string]$Message,
+
+        [Parameter(ParameterSetName = "pipeline", ValueFromPipeLine = $true, Position = 0)]
+        [Hashtable]$obj
+    )
+
+    Process {
```
Comment (rkeithhill) [inline src/WritePlasterContent.ps1]: lower case `process` here for consistency.

File: src/WritePlasterContent.ps1 [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+        # If a hash is passed then recall this function with the hash splatted instead.
+        if ($obj -ne $null) {
```
Comment (rkeithhill) [inline src/WritePlasterContent.ps1]: To shut up a PSScriptAnalyzer warning, change to `$null -ne $obj`. BTW I recommend you use Visual Studio Code to do Plaster development as it will mark the above line with a PSSA warning.

File: src/WritePlasterContent.ps1 [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+        # New Node
+        [System.XML.XMLElement]$oXMLContent = $oXMLRoot.CreateElement($Type)
+
+        # Append as child to an existing node
+        $null = $oXMLRoot.appendChild($oXMLContent)
+
+        # Update the element based on the type
+        switch ($Type) {
+          'File' {
+            <#
```
Comment (rkeithhill) [inline src/WritePlasterContent.ps1]: For inline comments, just use:
```posh
# Uses: source, destination, condition, openInEditor
```
Even for multiline comments
```posh
# comment line 1
# comment line 2
```

File: src/WritePlasterParameter.ps1 [hunk 13]
```diff
@@ -0,0 +1,87 @@
+function Write-PlasterParameter {
```
Comment (rkeithhill) [inline src/WritePlasterParameter.ps1]: Apply all relevant comments from previous function to this one.

File: src/WritePlasterContent.ps1 [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+            if (-not [string]::IsNullOrEmpty($message)) {
+                $oXMLContent.SetAttribute("message", $message)
+            }
+          }
+        }
+
+        $oXMLRoot.InnerXML
+
+    }
+}
```
Comment (rkeithhill) [inline src/WritePlasterContent.ps1]: Please add a newline at the end of the file.  I'm going to add that setting to the VSCode settings file - then it will take care of this issue.

File: src/WritePlasterParameter.ps1 [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+        $oXMLParameter.SetAttribute("type", $Type)
+        $oXMLParameter.SetAttribute("prompt", $Prompt)
+        if (-not [string]::IsNullOrEmpty($Default)) {
+            $oXMLParameter.SetAttribute("default", $Default)
+        }
+        if (-not [string]::IsNullOrEmpty($Store)) {
+            $oXMLParameter.SetAttribute("store", $Store)
+        }
+        if ($Type -eq 'choice') {
+            ForEach ($choice in $Choices) {
```
Comment (rkeithhill) [inline src/WritePlasterParameter.ps1]: lower case for keywords `foreach` as well like the `if` above it.  :-)

File: src/WritePlasterParameter.ps1 [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+                [System.XML.XMLElement]$oXMLChoice = $oXMLRoot.CreateElement("choice")
+                $oXMLChoice.SetAttribute("label", $choice.Label)
+                $oXMLChoice.SetAttribute("help", $choice.help)
+                $oXMLChoice.SetAttribute("value", $choice.value)
+                $null = $oXMLRoot['parameter'].appendChild($oXMLChoice)
+            }
+        }
+
+        $oXMLRoot.InnerXML
+
```
Comment (rkeithhill) [inline src/WritePlasterParameter.ps1]: Remove extra line.

File: src/WritePlasterContent.ps1 [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+                $oXMLContent.SetAttribute("requiredversion", $requiredversion)
+            }
+            if (-not [string]::IsNullOrEmpty($message)) {
+                $oXMLContent.SetAttribute("message", $message)
+            }
+          }
+        }
+
+        $oXMLRoot.InnerXML
+
```
Comment (rkeithhill) [inline src/WritePlasterContent.ps1]: Remove extra line.

**Comment by zloeber** [comment]
Wow, that was a quick turnaround and review of things. I honestly expected the pull request to sit for a while. Regardless, I made some mistakes on the write-plastercontent function that I needed to rectify anyway. I was questioning some of the PascalCase formatting as it didn't seem to be consistent within the actual schema file. (companyName, rootModule, author, et cetera).

I'm extremely grateful for the review and will make all necessary changes. Thanks for nudging me in the right direction instead of stomping all over me, it absolutely is encouraging and helpful and I appreciate it probably more than you might realize.

**Comment by rkeithhill** [comment]
No problem. It's great to have more contributors.  Oh, and ignore the comments about bumping the schema version.  Apparently I already updated the schema version for some changes to support Linux.  :-)

File: src/Schema/PlasterManifest-v1.xsd (hunk 4 above)
Comment (rkeithhill) [inline src/Schema/PlasterManifest-v1.xsd]: Ignore this.  As I mentioned, I had already bumped the schema version for other (unreleased) changes I made to the schema.

**Comment by daviwil** [comment]
Thanks @zloeber, awesome that you decided to contribute!

Yep, we should release Plaster 1.1 with the next PowerShell extension update since I need to make a change to Plaster's prompting behavior when run inside of the editor.  I'll try to get a PR for that sent tomorrow.

**Comment by zloeber** [comment]
I believe I've covered the bases for most of the items that have been pointed out. If you haven't figured it out I'm really just trying to lower the bar of entry to Plaster by making the parameter and content blocks a bit easier to fabricate than manually crafting xml. I'll take no offense if the proposed functions don't make the cut for being exported in the next release of the module (though I'll still use them to create my Plaster manifest files). Thanks again for tolerating my pull request from nowhere and giving constructive feedback. I don't contribute to projects very often so I'm a bit green on all this.

**Comment by rkeithhill** [comment]
@zloeber Thanks for making those changes.  I'll try to carve out some time tonight to take the helper commands for a test spin. 

BTW another way to ease creation of the Plaster template is to use the VSCode snippets we ship in the `snippets` dir of the Plaster module.  Unfortunately you have to manually copy these to your user snippets folder e.g. `C:\Users\Keith\AppData\Roaming\Code\User\snippets\xml.json`

**Comment by zloeber** [comment]
I was testing this a little further and the additional module manifest properties never get added. I tinkered a bit with some of the code in invoke-plaster to get it working but I think I'm still missing the logic point of pulling in just some of the module manifest properties if the manifest file already exists. Anyway, here are the changes I made (specifically to ProcessNewModuleManifest and CopyModuleManifestPropertyToHashtable). It makes the attribute processing a bit more generic but then introduces more exceptions (openineditor, encoding, destination, et cetera). And Tags still don't work...

Not doing another pull request on this as I've futzed with some of the core module manifest creation code so it's more of a consideration I suppose.

```
function CopyModuleManifestPropertyToHashtable([PSModuleInfo]$oldModuleManifest, [hashtable]$hashtable, [string[]]$Property) {
    foreach ($prop in $Property) {
        # if the property exists in the old manifest file but not in our existing hash then add it.
        if (($oldModuleManifest.$prop) -and ($null -eq $hashtable.$prop)) {
            if ($prop -eq 'Tags') {
                $hashtable[$prop] = @($oldModuleManifest.$prop)
            }
            else {
                $hashtable[$prop] = ($oldModuleManifest.$prop).ToString()
            }
        }
    }
}

function ProcessNewModuleManifest([ValidateNotNull()]$Node) {
    $DefinedProperties = @($node | Get-Member -Type:Property).Name
    $newModuleManifestParams = @{}
    ForEach ($ModProp in $DefinedProperties) {
        $PropVal = InterpolateAttributeValue $Node.$ModProp (GetErrorLocationNewModManifestAttrVal $ModProp)
        if (![string]::IsNullOrWhiteSpace($PropVal)) {
            switch ($ModProp) {
                'tags' {
                    $newModuleManifestParams['Tags'] = @($PropVal -split ',' | ForEach-Object {$_.trim()})
                }
                'destination' {
                    $dstRelPath = $PropVal
                }
                'encoding' {}
                'openInEditor' {}
                Default {
                    $newModuleManifestParams[$ModProp] = $PropVal
                }
            }
        }
    }

    # We could choose to not check this if the condition eval'd to false
    # but I think it is better to let the template author know they've broken the
    # rules for any of the file directives (not just the ones they're testing/enabled).
    if ([System.IO.Path]::IsPathRooted($dstRelPath)) {
        throw ($LocalizedData.ErrorPathMustBeRelativePath_F2 -f $dstRelPath,$Node.LocalName)
    }

    $dstPath = $PSCmdlet.GetUnresolvedProviderPathFromPSPath((Join-Path $DestinationPath $dstRelPath))

    $condition  = $Node.condition
    if ($condition -and !(EvaluateConditionAttribute $condition "'<$($Node.LocalName)>'")) {
        $PSCmdlet.WriteDebug("Skipping module manifest generation for '$dstPath', condition evaluated to false.")
        return
    }

    $encoding = $Node.encoding
    if (!$encoding) {
        $encoding = $DefaultEncoding
    }

    if ($PSCmdlet.ShouldProcess($dstPath, $LocalizedData.ShouldProcessNewModuleManifest)) {
        $manifestDir = Split-Path $dstPath -Parent
        if (!(Test-Path $manifestDir)) {
            VerifyPathIsUnderDestinationPath $manifestDir
            Write-Verbose ($LocalizedData.NewModManifest_CreatingDir_F1 -f $manifestDir)
            New-Item $manifestDir -ItemType Directory > $null
        }

        # If there is an existing module manifest, load it so we can reuse old values not specified by
        # template.
        if (Test-Path -LiteralPath $dstPath) {
            $oldModuleManifest = Test-ModuleManifest -Path $dstPath -ErrorAction SilentlyContinue
            if ($? -and $oldModuleManifest) {
                # We are only interested in module manifest properties not defined in the plaster manifest
                $props = 'Guid', 'Description', 'DefaultCommandPrefix', 'RootModule', 'AliasesToExport',
                            'CmdletsToExport', 'DscResourcesToExport', 'VariablesToExport',
                            'FormatsToProcess', 'TypesToProcess', 'ScriptsToProcess', 'Tags', 'Author',
                            'CompanyName', 'LicenseUri', 'ProjectUri', 'IconUri' | Where-Object {$DefinedProperties -notcontains $_}

                if ($props.count -gt 0) {
                    CopyModuleManifestPropertyToHashtable $oldModuleManifest $newModuleManifestParams $props
                }
            }
        }

        $tempFile = $null

        try {
            $tempFileBaseName = "moduleManifest-" + [Guid]::NewGuid()
            $tempFile = [System.IO.Path]::GetTempPath() + "${tempFileBaseName}.psd1"
            $PSCmdlet.WriteDebug("Created temp file for new module manifest - $tempFile")
            $newModuleManifestParams['Path'] = $tempFile

            # Generate manifest into a temp file.
            New-ModuleManifest @newModuleManifestParams

            # Typically the manifest is re-written with a new encoding (UTF8-NoBOM) because Git hates UTF-16.
            $content = Get-Content -LiteralPath $tempFile -Raw

            # Replace the temp filename in the generated manifest file's comment header with the actual filename.
            $dstBaseName = [System.IO.Path]::GetFileNameWithoutExtension($dstPath)
            $content = $content -replace "(?<=\s*#.*?)$tempFileBaseName", $dstBaseName

            WriteContentWithEncoding -Path $tempFile -Content $content -Encoding $encoding

            CopyFileWithConflictDetection $tempFile $dstPath

            if ($PassThru -and ($Node.openInEditor -eq 'true')) {
                $InvokePlasterInfo.OpenFiles += $dstPath
            }
        }
        finally {
            if ($tempFile -and (Test-Path $tempFile)) {
                Remove-Item -LiteralPath $tempFile
                $PSCmdlet.WriteDebug("Removed temp file for new module manifest - $tempFile")
            }
        }
    }
}
```

**Comment by zloeber** [comment]
The tags issue was a plaster manifest problem on my end. They do work. Here is one other alternate to allow for all valid new-modulemanifest parameters that are attainable through test-modulemanifest for reusing when the manifest file exist but values are not being passed through Plaster. This should hopefully cover for all new plaster schema changes for module manifest files moving forward (I think)
```
function CopyModuleManifestPropertyToHashtable([PSModuleInfo]$oldModuleManifest, [hashtable]$hashtable, [string[]]$Property) {
    foreach ($prop in $Property) {
        # if the property exists in the old manifest file but not in our existing hash then add it.
        if (($oldModuleManifest.$prop) -and ($null -eq $hashtable.$prop)) {
            if (($prop -eq 'Tags') -and (($oldModuleManifest.$prop).Count -gt 0)) {
                $hashtable[$prop] = @($oldModuleManifest.$prop)
            }
            else {
                $hashtable[$prop] = ($oldModuleManifest.$prop).ToString()
            }
        }
    }
}

function ProcessNewModuleManifest([ValidateNotNull()]$Node) {
    $DefinedProperties = @($node | Get-Member -Type:Property).Name
    $newModuleManifestParams = @{}
    ForEach ($ModProp in $DefinedProperties) {
        $PropVal = InterpolateAttributeValue $Node.$ModProp (GetErrorLocationNewModManifestAttrVal $ModProp)
        if (![string]::IsNullOrWhiteSpace($PropVal)) {
            switch ($ModProp) {
                'tags' {
                    Write-Output "Test Tags: $PropVal"
                    $newModuleManifestParams['Tags'] = @($PropVal -split ',' | ForEach-Object {$_.trim()})
                }
                'destination' {
                    $dstRelPath = $PropVal
                }
                'encoding' {}
                'openInEditor' {}
                Default {
                    $newModuleManifestParams[$ModProp] = $PropVal
                }
            }
        }
    }

    # We could choose to not check this if the condition eval'd to false
    # but I think it is better to let the template author know they've broken the
    # rules for any of the file directives (not just the ones they're testing/enabled).
    if ([System.IO.Path]::IsPathRooted($dstRelPath)) {
        throw ($LocalizedData.ErrorPathMustBeRelativePath_F2 -f $dstRelPath, $Node.LocalName)
    }

    $dstPath = $PSCmdlet.GetUnresolvedProviderPathFromPSPath((Join-Path $DestinationPath $dstRelPath))

    $condition = $Node.condition
    if ($condition -and !(EvaluateConditionAttribute $condition "'<$($Node.LocalName)>'")) {
        $PSCmdlet.WriteDebug("Skipping module manifest generation for '$dstPath', condition evaluated to false.")
        return
    }

    $encoding = $Node.encoding
    if (!$encoding) {
        $encoding = $DefaultEncoding
    }

    if ($PSCmdlet.ShouldProcess($dstPath, $LocalizedData.ShouldProcessNewModuleManifest)) {
        $manifestDir = Split-Path $dstPath -Parent
        if (!(Test-Path $manifestDir)) {
            VerifyPathIsUnderDestinationPath $manifestDir
            Write-Verbose ($LocalizedData.NewModManifest_CreatingDir_F1 -f $manifestDir)
            New-Item $manifestDir -ItemType Directory > $null
        }

        # If there is an existing module manifest, load it so we can reuse old values not specified by
        # template.
        if (Test-Path -LiteralPath $dstPath) {
            $oldModuleManifest = Test-ModuleManifest -Path $dstPath -ErrorAction SilentlyContinue
            if ($? -and $oldModuleManifest) {
                # First get all non-advanced function parameters that new-modulemanifest uses
                $IgnoredManifestVals = @('Path', 'PrivateData', 'PassThru')
                $ValidModuleManifestParams = (Get-Command New-ModuleManifest).Parameters.keys | ForEach-Object { if (($a.Parameters[$_].Attributes.TypeId.Name -notcontains 'AliasAttribute') -and ($IgnoredManifestVals -notcontains $_)) {$_} }

                # Get a list of properties that have values already from the old manifest but
                # are not defined in the plaster manifest and are also valid new-modulemanifest parameters
                $props = @(($oldModuleManifest | Get-member -Type 'Property' |
                            Where-Object {($null -ne $oldModuleManifest.($_.Name)) -and
                            ($_.Name -ne 'PSData') -and
                            (-not [string]::IsNullOrEmpty(($oldModuleManifest))) -and
                            ($ValidModuleManifestParams -contains $_.Name)}).Name |
                        Where-Object {$DefinedProperties -notcontains $_})
                if ($props.count -gt 0) {
                    CopyModuleManifestPropertyToHashtable $oldModuleManifest $newModuleManifestParams $props
                }
            }
        }

        $tempFile = $null

        try {
            $tempFileBaseName = "moduleManifest-" + [Guid]::NewGuid()
            $tempFile = [System.IO.Path]::GetTempPath() + "${tempFileBaseName}.psd1"
            $PSCmdlet.WriteDebug("Created temp file for new module manifest - $tempFile")
            $newModuleManifestParams['Path'] = $tempFile

            # Generate manifest into a temp file.
            New-ModuleManifest @newModuleManifestParams

            # Typically the manifest is re-written with a new encoding (UTF8-NoBOM) because Git hates UTF-16.
            $content = Get-Content -LiteralPath $tempFile -Raw

            # Replace the temp filename in the generated manifest file's comment header with the actual filename.
            $dstBaseName = [System.IO.Path]::GetFileNameWithoutExtension($dstPath)
            $content = $content -replace "(?<=\s*#.*?)$tempFileBaseName", $dstBaseName

            WriteContentWithEncoding -Path $tempFile -Content $content -Encoding $encoding

            CopyFileWithConflictDetection $tempFile $dstPath

            if ($PassThru -and ($Node.openInEditor -eq 'true')) {
                $InvokePlasterInfo.OpenFiles += $dstPath
            }
        }
        finally {
            if ($tempFile -and (Test-Path $tempFile)) {
                Remove-Item -LiteralPath $tempFile
                $PSCmdlet.WriteDebug("Removed temp file for new module manifest - $tempFile")
            }
        }
    }
}
```

**Comment by zloeber** [comment]
Sorry, I made multiple mistakes on what could be a good addition to this project due to lack of git etiquette and experience. I think I have a better grip on things now and will split this PR out into two separate PRs based on your current code base. 

BTW, the large number of modifications was due to VS code auto-code formatting after pasting a bunch of the code, whoops. Another lesson learned...
